### una ves mas importamos las librerias, como usaremos un Dataset que no es de juguete no importamos Datasets

siempre importa primero las mas pesadas

In [1]:
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

importamos el data set que se debe encontrar en la misma carpeta que el modelo

In [2]:
df = pd.read_csv('insurance.csv')

In [7]:
df

age     sex     bmi  children smoker     region      charges
0      19  female  27.900         0    yes  southwest  16884.92400
1      18    male  33.770         1     no  southeast   1725.55230
2      28    male  33.000         3     no  southeast   4449.46200
3      33    male  22.705         0     no  northwest  21984.47061
4      32    male  28.880         0     no  northwest   3866.85520
...   ...     ...     ...       ...    ...        ...          ...
1333   50    male  30.970         3     no  northwest  10600.54830
1334   18  female  31.920         0     no  northeast   2205.98080
1335   18  female  36.850         0     no  southeast   1629.83350
1336   21  female  25.800         0     no  southwest   2007.94500
1337   61  female  29.070         0    yes  northwest  29141.36030

[1338 rows x 7 columns]

hacemos un describe para ver si la tabla esta limpia

In [6]:
df.describe()

age          bmi     children       charges
count  1338.000000  1338.000000  1338.000000   1338.000000
mean     39.207025    30.663397     1.094918  13270.422265
std      14.049960     6.098187     1.205493  12110.011237
min      18.000000    15.960000     0.000000   1121.873900
25%      27.000000    26.296250     0.000000   4740.287150
50%      39.000000    30.400000     1.000000   9382.033000
75%      51.000000    34.693750     2.000000  16639.912515
max      64.000000    53.130000     5.000000  63770.428010

como podemos ver hay columnas que contienen texto por lo que no aparecen en el describe, ahora haremos un header para ver los datos y luego un unique

In [8]:
df.head()

age     sex     bmi  children smoker     region      charges
0   19  female  27.900         0    yes  southwest  16884.92400
1   18    male  33.770         1     no  southeast   1725.55230
2   28    male  33.000         3     no  southeast   4449.46200
3   33    male  22.705         0     no  northwest  21984.47061
4   32    male  28.880         0     no  northwest   3866.85520

In [9]:
df.region.unique()

array(['southwest', 'southeast', 'northwest', 'northeast'], dtype=object)

## ahora vamos a mapear las columnas no numericas para hacerlas numericas

In [10]:
df.sex = df.sex.map({'female' : 1, 'male' : 2})
df['smoker'] = df['smoker'].map({'no' : 0,'yes' : 1})
df.region = df.region.map({'southwest' : 1, 'southeast': 2, 'northwest': 3, 'northeast': 4})

ahora veremos los cambios

In [11]:
df.head()

age  sex     bmi  children  smoker  region      charges
0   19    1  27.900         0       1       1  16884.92400
1   18    2  33.770         1       0       2   1725.55230
2   28    2  33.000         3       0       2   4449.46200
3   33    2  22.705         0       0       3  21984.47061
4   32    2  28.880         0       0       3   3866.85520

### si el count de las columnas es menor al maximo, en este caso 1338, significa que tiene valores nulos, si mas del 50% son nulos se hace *dropna*, caso contrario un *fillna*

In [12]:
df.describe()

age          sex          bmi     children       smoker  \
count  1338.000000  1338.000000  1338.000000  1338.000000  1338.000000   
mean     39.207025     1.505232    30.663397     1.094918     0.204783   
std      14.049960     0.500160     6.098187     1.205493     0.403694   
min      18.000000     1.000000    15.960000     0.000000     0.000000   
25%      27.000000     1.000000    26.296250     0.000000     0.000000   
50%      39.000000     2.000000    30.400000     1.000000     0.000000   
75%      51.000000     2.000000    34.693750     2.000000     0.000000   
max      64.000000     2.000000    53.130000     5.000000     1.000000   

            region       charges  
count  1338.000000   1338.000000  
mean      2.484305  13270.422265  
std       1.104885  12110.011237  
min       1.000000   1121.873900  
25%       2.000000   4740.287150  
50%       2.000000   9382.033000  
75%       3.000000  16639.912515  
max       4.000000  63770.428010

tambien existe el ATRIBUTO details para familiarizarse con la dataframe

In [13]:
df.dtypes

age           int64
sex           int64
bmi         float64
children      int64
smoker        int64
region        int64
charges     float64
dtype: object

para elejir la columna target se tiene que elejir una columna NO CATEGORICA, las categoricas son aquellas que pueden dividirse en unos uantos valores finitos, mientras las no categoricas no pueden ser divididas de esa manera por ejemplo: 
### sexo (hombre, mujer) -- > (0,1) // categorica
### monto (1.....3999) --- > (1....3999) // no categorica

esta tabla sera nuestro objetivo y de la cual haremos predicciones usando otra columna

ahora dividiremos el dataset en un X (caracteristica independiente) y Y (caracteristica dependiente)

# si queremos sacar las columnas, elejimos las que tengan mas correlacion con la columna objetivo

In [15]:
corr = round(df.corr(),4)
corr.style.background_gradient()

# por ahora tomaremos la salida facil

In [14]:
data_x = df.drop(['charges'], axis=1)
data_y = df.charges

dividimos entre datos de entrenamiento y testeo

In [16]:
data_x_train,data_x_test,data_y_train,data_y_test =train_test_split(data_x,data_y,test_size=0.2)

## inicializamos la regresion lineal o el modelo

In [17]:
linear_reg = LinearRegression()

In [18]:
linear_reg.fit(data_x_train,data_y_train)

LinearRegression()

# veamos la ecuacion de la recta

In [20]:
linear_reg.coef_

array([  252.83233052,  -335.45011707,   343.63579016,   528.83612234,
       24063.78410571,   285.91409305])

In [21]:
linear_reg.intercept_

-12832.51310066256

# como ya entrenamos
vamos a hacer la prediccion

In [26]:
data_y_predicted = linear_reg.predict(data_x_test)

In [27]:
data_y_predicted

array([32436.04772017,  3453.19709555,  9093.93189846,  6610.11387195,
       32483.3563984 , 37165.82774842,  4235.30675618, 10678.64408138,
        6919.46332093,  7362.30429425, 14374.30728134,  7595.56206624,
       35942.5339224 ,  9209.34918078, 27975.65516396,  6383.79824899,
        4418.71554145, 14554.86126809,   761.86922483,  1630.47312341,
       10570.99222495, 11863.73288767,  1468.5919639 ,  9348.10110615,
       15229.79766194,  9741.77717701,  2240.81607978, 10474.17607979,
       34233.77068226, 13002.75296051,  3677.84932728, 10026.15960903,
       10836.41893011, 16561.11272321,  -900.96500502, 37788.57029558,
        8257.94400629, 27960.94202588, 12565.30283262,  6189.98871477,
       14652.27560678,  2875.26692427,  6007.26832851, 27725.80999535,
        3063.05951577,  8377.00459445, 14349.50694073, 27565.6230949 ,
       11371.37245047,  7104.93825413,  2652.22906518, 15964.25144267,
        9079.97505361, 13076.21757564,  6506.60424314, 39727.29272211,
      

ahora vemos el puntaje del resultado

In [30]:
print(f'Calificacion de Prediccion: {round(linear_reg.score(data_x_test,data_y_test),4)*100}%')

Calificacion de Prediccion: 73.9%


# ahora mejoramos la prediccion basandonos en la correlacion

In [32]:
new_data_x = df.drop(['sex','children','region','charges'],axis='columns')


In [38]:
new_data_x

age     bmi  smoker
0      19  27.900       1
1      18  33.770       0
2      28  33.000       0
3      33  22.705       0
4      32  28.880       0
...   ...     ...     ...
1333   50  30.970       0
1334   18  31.920       0
1335   18  36.850       0
1336   21  25.800       0
1337   61  29.070       1

[1338 rows x 3 columns]

In [57]:
new_data_x_train, new_data_x_test, y_train,y_test = train_test_split(new_data_x,data_y,test_size=0.2)

In [53]:
new_regression = LinearRegression()

In [58]:
new_regression.fit(new_data_x_train,y_train)

LinearRegression()

In [59]:
new_data_y_predicted = new_regression.predict(new_data_x_test)

In [60]:
print(f'Calificacion de Prediccion: {round(new_regression.score(new_data_x_test,y_test),4)*100}%')

Calificacion de Prediccion: 75.35%


ahora graficamos

In [64]:
def predict_charges(age,bmi,smoker,model):
    result = model.predict([[age,bmi,smoker]])
    return result[0]


In [71]:
age = 18
bmi = 32
smoker = 0

In [72]:
print(f'la persona de edad {age}, con indice de masa corporal {bmi} y que {"fuma" if smoker == 1 else "no fuma"}. Debe pagar: {predict_charges(age,bmi,smoker,new_regression)}')

la persona de edad 18, con indice de masa corporal 32 y que no fuma. Debe pagar: 3188.2476280917817


c:\Users\M-80-1\Documents\Trabajos de Clase\AlgoritmicaAvanzada\RepositorioAlgoritmica\algoritmica-2023-2\pandas_intro\pandas_env\lib\site-packages\sklearn\base.py:464: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(
